boa_engine/src/property/nonmaxu32.rs

```diff
@@ -0,0 +1,39 @@
+use std::num::NonZeroU32;
+
+/// An integer that is not `u32::MAX`.
+#[derive(PartialEq, Debug, Clone, Copy, Eq, Hash)]
+pub struct NonMaxU32 {
+    inner: NonZeroU32,
+}
+
+impl NonMaxU32 {
+    /// Creates a non-max `u32`.
+    ///
+    /// # Safety
+    ///
+    /// The caller must ensure that the given value is not `u32::MAX`.
+    #[must_use]
+    pub const unsafe fn new_unchecked(inner: u32) -> Self {
+        // SAFETY: The caller must ensure that `inner` is not `u32::MAX`.
+        let inner = unsafe { NonZeroU32::new_unchecked(inner + 1) };
```

Comment:
why add one and subtract when getting?

Reply:
Also, since this is an unsafe function and the user has to ensure the number is not `u32::MAX`, you can use `wrapping_add` in order to remove panicking code.

Reply:
> why add one and subtract when getting?

Since we are using `NonZeroU32` but our type excludes `u32::MAX` instead of `0`. If we would not do the addition we would run into `NonZeroU32::new_unchecked(0)` which is undefined behaviour.

Reply:
Ahh, Bu why not use a plain `u32`, `NonMaxU32` can be zero?

Reply:
There was a discussion over in https://github.com/boa-dev/boa/pull/2131 to use it for a potential non zero optimization. In my opinion we could also remove it and just use a `u32` since we only use `NonMaxU32` in `PropertyKey` which also has `JsString` and `JsSymbol` which are tagged pointers at `size = 8`. So optimizing would probably never work.

If we do that my only question is if we leave `NonMaxU32::new_unchecked` `unsafe`. It could not produce undefined behaviour, so in my view probably not, right?

Reply:
I see, thanks for the clarification :)

> If we do that my only question is if we leave `NonMaxU32::new_unchecked` `unsafe`. It could not produce undefined behaviour, so in my view probably not, right?

IMO, yes, We can leave a `debug_assert!` in it, just in case :)